src/ripple/unl/tests/BasicNetwork.h

```diff
@@ -317,55 +396,274 @@ class BasicNetwork
+    if (to)
+        by_to_[to].push_back(m);
+    if (from)
+        by_from_[from].push_back(m);
+    return by_when_.insert(m);
+}
+
+template <class Peer>
+void
+BasicNetwork<Peer>::queue_type::erase(
+    iterator iter)
+{
+    auto& m = *iter;
+    if (iter->to)
+    {
+        auto& list = by_to_[iter->to];
+        list.erase(list.iterator_to(m));
+    }
+    if (iter->from)
+    {
+        auto& list = by_from_[iter->from];
+        list.erase(list.iterator_to(m));
+    }
+    by_when_.erase(iter);
+    m.~msg();
+    alloc_.dealloc(&m, 1);
+}
+
+template <class Peer>
+void
+BasicNetwork<Peer>::queue_type::remove(
+    Peer* from, Peer* to)
+{
+    {
+        auto& list = by_to_[to];
+        for(auto iter = list.begin();
+            iter != list.end();)
+        {
+            if (iter->from == from)
+            {
+                auto& m = *iter;
+                iter = list.erase(iter);
+                m.~msg();
+                alloc_.dealloc(&m, 1);
+            }
+            else
+            {
+                ++iter;
+            }
+        }
+    }
+    {
+        auto& list = by_from_[from];
+        for(auto iter = list.begin();
+            iter != list.end();)
+        {
+            if (iter->to == to)
+            {
+                auto& m = *iter;
+                iter = list.erase(iter);
+                m.~msg();
+                alloc_.dealloc(&m, 1);
+            }
+            else
+            {
+                ++iter;
+            }
+        }
+    }
+}
 
 //------------------------------------------------------------------------------
 
-template <class Peer, class Derived>
-template <class Message>
-class BasicNetwork<Peer,Derived>::invoke_impl
-    : public invoke
+template <class Peer>
+class BasicNetwork<Peer>::link_transform
 {
+private:
+    BasicNetwork& net_;
+    Peer& from_;
+
 public:
-    invoke_impl (Derived& net,
-            Peer& peer, Message&& m)
-        : peer_(peer)
-        , net_(net)
-        , m_(std::forward<Message>(m))
+    using argument_type =
+        typename links_type::value_type;
+
+    class result_type
+    {
+    public:
+        Peer& to;
+        bool inbound;
+
+        result_type (result_type const&) = default;
+
+        result_type (BasicNetwork& net, Peer& from,
+                Peer& to_, bool inbound_)
+            : to(to_)
+            , inbound(inbound_)
+            , net_(net)
+            , from_(from)
+        {
+        }
+
+        /** Disconnect this link.
+
+            Effects:
+
+                The connection is removed at both ends.
+        */
+        void
+        disconnect() const
+        {
+            net_.disconnect(from_, to);
+        }
+
+    private:
+        BasicNetwork& net_;
+        Peer& from_;
+    };
+
+    link_transform (BasicNetwork& net, Peer& from)
+        : net_(net)
+        , from_(from)
     {
     }
 
-    void
-    operator()(Peer& from) const override
+    result_type const
```

Comment:
Yes, this is needed because of `boost::transform_iterator`, or else compilation fails in gcc and clang.